test/internal/pbxproj_partials/write_target_build_settings_tests.bzl

```
@@ -0,0 +1,796 @@
+"""Tests for `pbxproj_partials.write_target_build_settings`."""
+
+load("@bazel_skylib//lib:unittest.bzl", "asserts", "unittest")
+load("//test:mock_actions.bzl", "mock_actions")
+
+# buildifier: disable=bzl-visibility
+load("//xcodeproj/internal:pbxproj_partials.bzl", "pbxproj_partials")
+
+_BUILD_SETTINGS_DECLARED_FILE = mock_actions.mock_file(
+    "a_target_name.rules_xcodeproj.build_settings",
+)
+_DEBUG_SETTINGS_DECLARED_FILE = mock_actions.mock_file(
+    "a_target_name.rules_xcodeproj.debug_settings",
+)
+_C_PARAMS_DECLARED_FILE = mock_actions.mock_file(
+    "a_target_name.c.compile.params",
+)
+_CXX_PARAMS_DECLARED_FILE = mock_actions.mock_file(
+    "a_target_name.cxx.compile.params",
+)
+
+def _write_target_build_settings_test_impl(ctx):
+    env = unittest.begin(ctx)
+
+    # Arrange
+
+    actions = mock_actions.create()
+
+    expected_declared_files = {}
+    expected_outputs = []
+    expected_params = []
+
+    if ctx.attr.expect_build_settings:
+        expected_build_settings = _BUILD_SETTINGS_DECLARED_FILE
+        expected_declared_files[expected_build_settings] = None
+        expected_outputs.append(expected_build_settings)
+    else:
+        expected_build_settings = None
+
+    swift_debug_settings_to_merge = depset(
+        ctx.attr.swift_debug_settings_to_merge,
+    )
+
+    if ctx.attr.expect_debug_settings:
+        expected_debug_settings = _DEBUG_SETTINGS_DECLARED_FILE
+        expected_declared_files[expected_debug_settings] = None
+        expected_inputs = swift_debug_settings_to_merge
+        expected_outputs.append(expected_debug_settings)
+    else:
+        expected_debug_settings = None
+        expected_inputs = []
```

[nit] Just to match the `expected_outputs` pattern above:
```suggestion
    expected_inputs = []

    if ctx.attr.expect_debug_settings:
        expected_debug_settings = _DEBUG_SETTINGS_DECLARED_FILE
        expected_declared_files[expected_debug_settings] = None
        expected_inputs = swift_debug_settings_to_merge
        expected_outputs.append(expected_debug_settings)
    else:
        expected_debug_settings = None
```

---

I like my way in performance critical code. I'll go with your way for readability.